mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,571 @@
+  }
+  return success();
+}
+
+static std::optional<std::tuple<int64_t, int64_t, ArrayAttr>>
+getIdToLDS(ThreadwiseWriteAllOp &op, OpBuilder &b) {
+  ArrayAttr srcTransform = op.getExtraViewsAttr();
+  SetVector<StringRef> dimensionsToRemove;
+  dimensionsToRemove.insert("g_block");
+  dimensionsToRemove.insert("m_block");
+  dimensionsToRemove.insert("n_block");
+  // for attention:
+  dimensionsToRemove.insert("gblock");
+  dimensionsToRemove.insert("mblock");
+  dimensionsToRemove.insert("nblock");
+  FailureOr<ArrayAttr> maybeIdToLDS =
+      removeUpperDims(b, srcTransform, dimensionsToRemove);
+  if (failed(maybeIdToLDS)) {
+    LLVM_DEBUG(llvm::dbgs() << "getIdToLDS failed\n");
+    return std::nullopt;
+  }
+  ArrayAttr idToLDS = maybeIdToLDS.value();
+
+  ArrayRef<int64_t> shape = getLowerShape(idToLDS);
+  if (shape.size() != 2) {
+    LLVM_DEBUG(llvm::dbgs() << "getIdToLDS failed\n");
+    return std::nullopt;
+  }
+  int64_t mPerBlock = shape[0];
+  int64_t nPerBlock = shape[1];
+
+  return std::make_tuple(mPerBlock, nPerBlock, idToLDS);
+}
+
+struct ThreadwiseWriteAllRewritePattern
+    : public OpRewritePattern<ThreadwiseWriteAllOp> {
+  using OpRewritePattern<ThreadwiseWriteAllOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ThreadwiseWriteAllOp op,
+                                PatternRewriter &b) const override {
+    Location loc = op.getLoc();
+
+    // Prepare some useful constants.
+    Value convertedC = op.getSource();
+    Value matC = op.getDest();
+    Type destType = op.getDest().getType().getElementType();
+
+    // Convert from reg -> memory transform to reg -> block
+    int64_t mPerBlock, nPerBlock;
+    ArrayAttr idToLDS;
+    std::optional<std::tuple<int64_t, int64_t, ArrayAttr>> maybeBlockInfo =
+        getIdToLDS(op, b);
+    if (!maybeBlockInfo.has_value()) {
+      return failure();
+    }
+    std::tie(mPerBlock, nPerBlock, idToLDS) = maybeBlockInfo.value();
+
+    // Obtain critical matrix dimensions.
+    ArrayRef<int64_t> cShape;
+    cShape = op.getDest().getType().getShape();
+    if (cShape.size() != 3) {
+      return failure();
+    }
+    int64_t G = cShape[0];
+    int64_t M = cShape[1];
+    int64_t N = cShape[2];
+    LLVM_DEBUG(llvm::dbgs() << "G: " << G << " N: " << N << " M: " << M
+                            << " nPerBlock: " << nPerBlock
+                            << " mPerBlock: " << mPerBlock << "\n");
+    FailureOr<IntegerAttr> maybeGridSize = getGridSize(op);
+    if (failed(maybeGridSize)) {
+      return failure();
+    }
+    FailureOr<IntegerAttr> maybeBlockSize = getBlockSize(op);
+    if (failed(maybeBlockSize)) {
+      return failure();
+    }
+    int64_t blockSize = maybeBlockSize.value().getValue().getSExtValue();
+    int64_t gridSize = maybeGridSize.value().getValue().getSExtValue();
+    bool useIndexDiffs = true;
+    bool forceUnroll = true;
+    int64_t ldsRequiredBytes = mPerBlock * nPerBlock * getByteWidth(destType);
```

Comment:
This is not the right way to get this. You should already be getting this properly somewhere in "reuseLDS" code.
This is certainly not right for attention.

Reply:
i.e. walk the code and fetch the allocs

Reply:
I figure this is for checking if the pass should be applied?